File: zig/perfect-numbers/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

fn aliquot(n: u64) u64 {
    var s: u64 = 0;
    for (1..n) |i| {
        if (n % i == 0) s += i;
    }
    return s;
}

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    const sum: u64 = aliquot(n);
    if (n == sum) return .perfect;
    if (n < sum) return .abundant;
    return .deficient;
}

File: cpp/perfect-numbers/perfect_numbers.cpp
#include "perfect_numbers.h"

#include <stdexcept>

namespace perfect_numbers {
classification classify(int num) {
    if (num <= 0) throw std::domain_error("invalid input");

    int sum = 0;
    for (int i = 1; i < num; i++) {
        if (num % i == 0) {
            sum += i;
        }
    }

    if (num == sum)
        return classification::perfect;
    else if (num < sum)
        return classification::abundant;
    else
        return classification::deficient;
}
}  // namespace perfect_numbers
